In [2]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt   

In [5]:
file_path = "data\\train.csv"

df = pd.read_csv(file_path)

In [7]:
df.head()

carID   Brand      model    year  price transmission  mileage fuelType  \
0  69512      VW       Golf  2016.0  22290    Semi-Auto  28421.0   Petrol   
1  53000  Toyota      Yaris  2019.0  13790       Manual   4589.0   Petrol   
2   6366    Audi         Q2  2019.0  24990    Semi-Auto   3624.0   Petrol   
3  29021    Ford     FIESTA  2018.0  12500        anual   9102.0   Petrol   
4  10062     BMW   2 Series  2019.0  22995       Manual   1000.0   Petrol   

     tax        mpg  engineSize  paintQuality%  previousOwners  hasDamage  
0    NaN  11.417268         2.0           63.0        4.000000        0.0  
1  145.0  47.900000         1.5           50.0        1.000000        0.0  
2  145.0  40.900000         1.5           56.0        4.000000        0.0  
3  145.0  65.700000         1.0           50.0       -2.340306        0.0  
4  145.0  42.800000         1.5           97.0        3.000000        0.0

In [14]:
df.shape

(75973, 14)

In [11]:
df.columns

Index(['carID', 'Brand', 'model', 'year', 'price', 'transmission', 'mileage',
       'fuelType', 'tax', 'mpg', 'engineSize', 'paintQuality%',
       'previousOwners', 'hasDamage'],
      dtype='object')

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75973 entries, 0 to 75972
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   carID           75973 non-null  int64  
 1   Brand           74452 non-null  object 
 2   model           74456 non-null  object 
 3   year            74482 non-null  float64
 4   price           75973 non-null  int64  
 5   transmission    74451 non-null  object 
 6   mileage         74510 non-null  float64
 7   fuelType        74462 non-null  object 
 8   tax             68069 non-null  float64
 9   mpg             68047 non-null  float64
 10  engineSize      74457 non-null  float64
 11  paintQuality%   74449 non-null  float64
 12  previousOwners  74423 non-null  float64
 13  hasDamage       74425 non-null  float64
dtypes: float64(8), int64(2), object(4)
memory usage: 8.1+ MB


In [21]:
round(df.describe().T, 2)

count      mean       std       min       25%      50%  \
carID           75973.0  37986.00  21931.66      0.00  18993.00  37986.0   
year            74482.0   2017.10      2.21   1970.00   2016.00   2017.0   
price           75973.0  16881.89   9736.93    450.00  10200.00  14699.0   
mileage         74510.0  23004.18  22129.79 -58540.57   7423.25  17300.0   
tax             68069.0    120.33     65.52    -91.12    125.00    145.0   
mpg             68047.0     55.15     16.50    -43.42     46.30     54.3   
engineSize      74457.0      1.66      0.57     -0.10      1.20      1.6   
paintQuality%   74449.0     64.59     21.02      1.64     47.00     65.0   
previousOwners  74423.0      1.99      1.47     -2.35      1.00      2.0   
hasDamage       74425.0      0.00      0.00      0.00      0.00      0.0   

                    75%        max  
carID           56979.0   75972.00  
year             2019.0    2024.12  
price           20950.0  159999.00  
mileage         32427.5  323000.00  
tax               145.0     580.00  
mpg                62.8     470.80  
engineSize          2.0       6.60  
paintQuality%      82.0     125.59  
previousOwners      3.0       6.26  
hasDamage           0.0       0.00

In [30]:
df.describe(include="object").T

count unique     top   freq
Brand         74452     72    Ford  14808
model         74456    735   Focus   6353
transmission  74451     40  Manual  38050
fuelType      74462     34  Petrol  37995


- **COL year**
  - Diistribuição concentrada em torno da média. No enunciado é dito a bd é relativa ao ano de 2020 no entanto o valor maximo é 2024?
  - Faz sentido ter modelos depois de 2020?
  - Cerca de 1500 valores nulos.
  - Tranformar em int.

- **COL price**
  - Distribuição extremamente dispersa. Valor máximo muito alto a comparar com valor médio e quartis, indicativo da existencia de outliers.

- **COL mileage**
  - Distribuição extremamente dispersa. Valor do desvio padrão é praticamente igual ao da média. Novamente o valor maximo parece indicar a existência de outliers.
  - Valores negativos??
  - Cerca de 1500 valores nulos.

- **COL tax**
  - Cerca de 7000 valore nulos.
  - Valores negativos?? Necessario contextualização. Numa primeira analise não faz muito sentido dado que o valor está expresso em libras.

- **COL mpg**
  - Cerca de 7000 valore nulos.
  - Valores negativos?? Necessario contextualização. Numa primeira analise não faz muito sentido dado que o valor está expresso em Gallon (3.78541178 L).
  - Meter coluna em Litros.

- **COL engineSize**
  - Valores negativos?? Coluna em dm³ = L.
  - GPT: A cilindrada do motor (ou engine size) mede o volume total dos cilindros de um motor — ou seja, o espaço por onde o combustível e o ar são misturados e queimados.
  - Vamos esclarecer as unidades:
    - 1 litro (L) = 1 decímetro cúbico (dm³) = 1.000 centímetros cúbicos (cm³)
  - Portanto:
    - Um motor 1.0 L tem 1.000 cm³ de cilindrada.
    - Um motor 1.6 L tem 1.600 cm³.
    - Um motor 2.0 L tem 2.000 cm³.

- **COL paintQuality%**
  - Cerca de 1500 valores nulos.
  - MAX > 100% ?? Não faz muito sentido.

- **COL previousOwners**
  - Cerca de 1500 valores nulos.
  - Valores negativos??
  - Tranformar em int.

- **COL hasDamage**
  - Coluna boolean, mas todos os valores são 0.
  - Redundante 

- **Col year**
  - Variável categórica ordinal
  - Converter de float -> int (arredondar para cima ou baixo)

- **Col mileage**
  - Garantir que mileage > 0 (Nan ou 0)
  - Converter de gallons -> litros (L)

- **Col tax**
  - Verificar se tax > 0 (necessário conhecimento de negócio)

- **Col engineSize**
  - Garantir que engineSize > 0 (Nan ou 0)

- **Col paintQuality%**
  - Garantir que paintQuality% <= 100% (100% ou Nan)

- **Col previousOwners**
  - Variável categórica ordinal
  - Converter de float -> int (arredondar para cima ou baixo)

- **Col hasDamage**
  - Redundante
  - Remover coluna

- **Valores nulos**
  - Avaliar e tratar valores faltantes em cada coluna

In [35]:
sum(df["carID"].duplicated())

0

In [62]:
df["previousOwners"].unique()

array([ 4.        ,  1.        , -2.34030622,  3.        ,  0.        ,
        2.        ,         nan, -2.34565   ,  6.25837066, -2.33512284,
        6.22789796,  6.25823052,  6.23017958, -2.33936045,  6.21772443,
        6.24583495, -2.29943868, -2.34010209, -2.31225953,  6.2482512 ,
       -2.31733109,  6.23308217,  6.24177863, -2.33744529])

In [63]:
df["year"].unique()

array([2016.        , 2019.        , 2018.        , 2014.        ,
       2017.        , 2020.        , 2013.        ,           nan,
       2015.        , 2023.36707842, 2011.        , 2012.        ,
       2023.1169636 , 2003.        , 2009.        , 2007.        ,
       2005.        , 2011.11118842, 2011.21085349, 2023.97731126,
       2004.        , 2010.        , 2010.56500919, 2008.        ,
       2024.12175905, 2006.        , 2023.60527574, 2023.38982198,
       2010.67696784, 2001.        , 2000.        , 2023.26798867,
       2010.26863473, 2010.37154646, 2002.        , 2022.69668507,
       2009.81675711, 2010.7464032 , 2012.69574039, 1996.        ,
       1998.        , 2022.87800554, 1970.        , 1999.        ,
       1997.        ])

In [ ]:
col_c = df.columns[(df.dtypes == "object") | (df.nunique() < 20)].tolist()
col_n = df.columns.difference(col_c).tolist()

In [ ]:
col_n

['carID',
 'engineSize',
 'mileage',
 'mpg',
 'paintQuality%',
 'previousOwners',
 'price',
 'tax',
 'year']

In [47]:
col_c

['Brand', 'model', 'transmission', 'fuelType', 'hasDamage']

In [70]:
unique_values = df[col_c].apply(lambda x: x.unique()).to_dict()

for col, values in unique_values.items():
    print(f"{col}:\n{values}\n")

Brand:
['VW' 'Toyota' 'Audi' 'Ford' 'BMW' 'Skoda' 'Opel' 'Mercedes' 'FOR'
 'mercedes' 'Hyundai' 'w' 'ord' 'MW' 'bmw' nan 'yundai' 'BM' 'Toyot' 'udi'
 'Ope' 'AUDI' 'V' 'opel' 'pel' 'For' 'pe' 'Mercede' 'audi' 'MERCEDES'
 'OPEL' 'koda' 'FORD' 'Hyunda' 'W' 'Aud' 'vw' 'hyundai' 'skoda' 'ford'
 'TOYOTA' 'ercedes' 'oyota' 'toyota' 'SKODA' 'Skod' 'HYUNDAI' 'kod' 'v'
 'for' 'SKOD' 'aud' 'KODA' 'PEL' 'yunda' 'or' 'UDI' 'OYOTA' 'HYUNDA' 'mw'
 'OPE' 'mercede' 'ERCEDES' 'ercede' 'TOYOT' 'MERCEDE' 'ORD' 'ud' 'ope'
 'AUD' 'hyunda' 'skod' 'toyot']

model:
[' Golf' ' Yaris' ' Q2' ' FIESTA' ' 2 Series' '3 Series' ' A3' ' Octavia'
 ' Passat' ' Focus' ' Insignia' ' A Clas' ' Q3' ' Fabia' ' A Class' ' Ka+'
 ' 3 Series' ' GLC Class' ' I30' ' C Class' ' Polo' ' E Class' ' C Clas'
 ' Q5' ' Up' ' Fiesta' ' C-HR' ' Mokka X' ' Corsa' ' Astra' ' TT'
 ' 5 Series' ' Aygo' ' 4 Series' ' SLK' ' Viva' ' T-Roc' 'Focus'
 ' EcoSport' ' Tucson' ' EcoSpor' nan ' X-CLASS' ' CL Class' ' IX20'
 ' i20' ' Rapid' ' a1' ' Auris'

Categoricas Formatação:
    Aplicavel a todas as colunas:
    - Formatar tudo para minusculas ou no mesmo formato. Coerencia.
    - Valores com erros de inserção. Ou falta a primeira letra ou a ultima ou apenas consta parte da str. Pode ser resolvido com Substr()
    - Valores Incorretos? Apenas uma letra. "or", "w", "v", ...? 
    - nan